// @activepieces/piece-schedule@0.1.5/src/lib/triggers/cron-expression.trigger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cronExpressionTrigger = void 0;
const tslib_1 = require("tslib");
const pieces_framework_1 = require("@activepieces/pieces-framework");
const pieces_framework_2 = require("@activepieces/pieces-framework");
const common_1 = require("../common");
exports.cronExpressionTrigger = (0, pieces_framework_2.createTrigger)({
    name: 'cron_expression',
    displayName: 'Cron Expression',
    description: 'Trigger based on cron expression',
    props: {
        cronExpression: pieces_framework_2.Property.ShortText({
            displayName: 'Cron Expression',
            description: 'Cron expression to trigger',
            required: true,
            defaultValue: '0/5 * * * *',
        }),
        timezone: pieces_framework_2.Property.StaticDropdown({
            displayName: 'Timezone',
            options: {
                options: common_1.timezoneOptions,
            },
            required: true,
            defaultValue: 'UTC',
        }),
    },
    type: pieces_framework_1.TriggerStrategy.POLLING,
    sampleData: {},
    onEnable: (ctx) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        ctx.setSchedule({
            cronExpression: ctx.propsValue.cronExpression,
            timezone: ctx.propsValue.timezone,
        });
    }),
    run(context) {
        return Promise.resolve([{}]);
    },
    onDisable: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        console.log('onDisable');
    }),
});
