// @activepieces/piece-schedule@0.1.5/src/lib/triggers/every-month.trigger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.everyMonthTrigger = void 0;
const tslib_1 = require("tslib");
const pieces_framework_1 = require("@activepieces/pieces-framework");
const pieces_framework_2 = require("@activepieces/pieces-framework");
const common_1 = require("../common");
exports.everyMonthTrigger = (0, pieces_framework_2.createTrigger)({
    name: 'every_month',
    displayName: 'Every Month',
    description: 'Triggers the current flow every month',
    type: pieces_framework_1.TriggerStrategy.POLLING,
    sampleData: {},
    props: {
        day_of_the_month: pieces_framework_2.Property.StaticDropdown({
            displayName: 'Day of the month',
            options: {
                options: common_1.MONTH_DAYS.map((d, idx) => {
                    return {
                        label: (1 + d).toString(),
                        value: idx + 1,
                    };
                }),
            },
            required: true,
        }),
        hour_of_the_day: pieces_framework_2.Property.StaticDropdown({
            displayName: 'Hour of the day',
            options: {
                options: common_1.DAY_HOURS.map((d, idx) => {
                    return {
                        label: d,
                        value: idx,
                    };
                }),
            },
            required: true,
        }),
        timezone: pieces_framework_2.Property.StaticDropdown({
            displayName: 'Timezone',
            options: {
                options: common_1.timezoneOptions,
            },
            required: true,
            defaultValue: 'UTC',
        }),
    },
    onEnable: (ctx) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        const hourOfTheDay = (0, common_1.validateHours)(ctx.propsValue.hour_of_the_day);
        const dayOfTheMonth = (0, common_1.validateMonthDays)(ctx.propsValue.day_of_the_month);
        const cronExpression = `0 ${hourOfTheDay} ${dayOfTheMonth} * *`;
        ctx.setSchedule({
            cronExpression: cronExpression,
            timezone: ctx.propsValue.timezone,
        });
    }),
    run(ctx) {
        const hourOfTheDay = (0, common_1.validateHours)(ctx.propsValue.hour_of_the_day);
        const dayOfTheMonth = (0, common_1.validateMonthDays)(ctx.propsValue.day_of_the_month);
        const cronExpression = `0 ${hourOfTheDay} ${dayOfTheMonth} * *`;
        return Promise.resolve([
            {
                hour_of_the_day: hourOfTheDay,
                day_of_the_month: dayOfTheMonth,
                cron_expression: cronExpression,
                timezone: ctx.propsValue.timezone,
            },
        ]);
    },
    onDisable: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        console.log('onDisable');
    }),
});
