// @activepieces/piece-schedule@0.1.5/src/lib/triggers/every-x-minutes.trigger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.everyXMinutesTrigger = void 0;
const tslib_1 = require("tslib");
const pieces_framework_1 = require("@activepieces/pieces-framework");
exports.everyXMinutesTrigger = (0, pieces_framework_1.createTrigger)({
    name: 'every_x_minutes',
    displayName: 'Every X Minutes',
    description: 'Triggers the current flow every X minutes',
    type: pieces_framework_1.TriggerStrategy.POLLING,
    sampleData: {},
    props: {
        minutes: pieces_framework_1.Property.StaticDropdown({
            displayName: 'Minutes',
            description: 'Valid value between 1 to 59.',
            required: true,
            defaultValue: 1,
            options: {
                disabled: false,
                options: Array.from({ length: 59 }, (_, index) => ({
                    label: `${index + 1} minute${index !== 0 ? 's' : ''}`,
                    value: index + 1,
                })),
            },
        }),
    },
    onEnable: (ctx) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        const cronExpression = `*/${ctx.propsValue.minutes} * * * *`;
        ctx.setSchedule({
            cronExpression: cronExpression,
            timezone: 'UTC',
        });
    }),
    run(ctx) {
        const cronExpression = `*/${ctx.propsValue.minutes} * * * *`;
        return Promise.resolve([
            {
                cron_expression: cronExpression,
                timezone: 'UTC',
            },
        ]);
    },
    onDisable: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        console.log('onDisable');
    }),
});
