// @activepieces/piece-schedule@0.1.5/src/lib/triggers/every-hour.trigger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.everyHourTrigger = void 0;
const tslib_1 = require("tslib");
const pieces_framework_1 = require("@activepieces/pieces-framework");
const pieces_framework_2 = require("@activepieces/pieces-framework");
exports.everyHourTrigger = (0, pieces_framework_2.createTrigger)({
    name: 'every_hour',
    displayName: 'Every Hour',
    description: 'Triggers the current flow every hour',
    type: pieces_framework_1.TriggerStrategy.POLLING,
    sampleData: {},
    props: {
        run_on_weekends: pieces_framework_2.Property.Checkbox({
            displayName: 'Run on weekends (Sat,Sun)',
            required: true,
            defaultValue: false,
        }),
    },
    onEnable: (ctx) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        const cronExpression = ctx.propsValue.run_on_weekends
            ? `0 * * * *`
            : `0 * * * 1-5`;
        ctx.setSchedule({
            cronExpression: cronExpression,
            timezone: 'UTC',
        });
    }),
    run(ctx) {
        const cronExpression = ctx.propsValue.run_on_weekends
            ? `0 * * * *`
            : `0 * * * 1-5`;
        return Promise.resolve([
            {
                cron_expression: cronExpression,
                timezone: 'UTC',
            },
        ]);
    },
    onDisable: () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        console.log('onDisable');
    }),
});
