// @activepieces/shared@0.10.107/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RetryFlowRequestBody = exports.STORE_KEY_MAX_LENGTH = exports.WebhookPauseMetadata = exports.PauseMetadata = exports.DelayPauseMetadata = exports.FlowRunResponse = exports.FlowRunStatus = exports.PauseType = exports.StopResponse = exports.FlowError = exports.GenerateCodeResponse = exports.GenerateCodeRequest = exports.isFlowStateTerminal = exports.FileResponseInterface = exports.secureApId = exports.ApId = exports.apId = exports.SeekPage = exports.FlowRetryStrategy = exports.RunEnvironment = exports.FlowRun = exports.Flow = exports.FlowVersionMetadata = exports.FlowVersionState = exports.FlowVersion = exports.AUTHENTICATION_PROPERTY_NAME = exports.TriggerType = exports.PieceTrigger = exports.PieceTriggerSettings = exports.EmptyTrigger = exports.Trigger = exports.TestFlowRunRequestBody = exports.SignInRequest = exports.SignUpRequest = exports.ExecuteCodeRequest = exports.StepOutputStatus = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./lib/flows/actions/action"), exports);
tslib_1.__exportStar(require("./lib/app-connection/app-connection"), exports);
tslib_1.__exportStar(require("./lib/app-connection/dto/read-app-connection-request"), exports);
tslib_1.__exportStar(require("./lib/app-connection/dto/upsert-app-connection-request"), exports);
tslib_1.__exportStar(require("./lib/common"), exports);
tslib_1.__exportStar(require("./lib/common/activepieces-error"), exports);
tslib_1.__exportStar(require("./lib/common/telemetry"), exports);
tslib_1.__exportStar(require("./lib/engine"), exports);
tslib_1.__exportStar(require("./lib/flag/flag"), exports);
tslib_1.__exportStar(require("./lib/flow-run/dto/list-flow-runs-request"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/execution-output"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/step-output"), exports);
tslib_1.__exportStar(require("./lib/flows/flow-operations"), exports);
tslib_1.__exportStar(require("./lib/flows/step-run"), exports);
tslib_1.__exportStar(require("./lib/flow-run/execution/execution-output"), exports);
var step_output_1 = require("./lib/flow-run/execution/step-output");
Object.defineProperty(exports, "StepOutputStatus", { enumerable: true, get: function () { return step_output_1.StepOutputStatus; } });
tslib_1.__exportStar(require("./lib/pieces"), exports);
tslib_1.__exportStar(require("./lib/store-entry/dto/store-entry-request"), exports);
tslib_1.__exportStar(require("./lib/webhook"), exports);
tslib_1.__exportStar(require("./lib/flows/dto/count-flows-request"), exports);
var code_request_1 = require("./lib/code/dto/code-request");
Object.defineProperty(exports, "ExecuteCodeRequest", { enumerable: true, get: function () { return code_request_1.ExecuteCodeRequest; } });
tslib_1.__exportStar(require("./lib/authentication/dto/authentication-response"), exports);
var sign_up_request_1 = require("./lib/authentication/dto/sign-up-request");
Object.defineProperty(exports, "SignUpRequest", { enumerable: true, get: function () { return sign_up_request_1.SignUpRequest; } });
var sign_in_request_1 = require("./lib/authentication/dto/sign-in-request");
Object.defineProperty(exports, "SignInRequest", { enumerable: true, get: function () { return sign_in_request_1.SignInRequest; } });
tslib_1.__exportStar(require("./lib/authentication/model/principal-type"), exports);
tslib_1.__exportStar(require("./lib/flows/actions/action"), exports);
tslib_1.__exportStar(require("./lib/user"), exports);
var test_flow_run_request_1 = require("./lib/flow-run/test-flow-run-request");
Object.defineProperty(exports, "TestFlowRunRequestBody", { enumerable: true, get: function () { return test_flow_run_request_1.TestFlowRunRequestBody; } });
var trigger_1 = require("./lib/flows/triggers/trigger");
Object.defineProperty(exports, "Trigger", { enumerable: true, get: function () { return trigger_1.Trigger; } });
Object.defineProperty(exports, "EmptyTrigger", { enumerable: true, get: function () { return trigger_1.EmptyTrigger; } });
Object.defineProperty(exports, "PieceTriggerSettings", { enumerable: true, get: function () { return trigger_1.PieceTriggerSettings; } });
Object.defineProperty(exports, "PieceTrigger", { enumerable: true, get: function () { return trigger_1.PieceTrigger; } });
Object.defineProperty(exports, "TriggerType", { enumerable: true, get: function () { return trigger_1.TriggerType; } });
Object.defineProperty(exports, "AUTHENTICATION_PROPERTY_NAME", { enumerable: true, get: function () { return trigger_1.AUTHENTICATION_PROPERTY_NAME; } });
var flow_version_1 = require("./lib/flows/flow-version");
Object.defineProperty(exports, "FlowVersion", { enumerable: true, get: function () { return flow_version_1.FlowVersion; } });
Object.defineProperty(exports, "FlowVersionState", { enumerable: true, get: function () { return flow_version_1.FlowVersionState; } });
Object.defineProperty(exports, "FlowVersionMetadata", { enumerable: true, get: function () { return flow_version_1.FlowVersionMetadata; } });
var flow_1 = require("./lib/flows/flow");
Object.defineProperty(exports, "Flow", { enumerable: true, get: function () { return flow_1.Flow; } });
tslib_1.__exportStar(require("./lib/file"), exports);
tslib_1.__exportStar(require("./lib/flows/flow-helper"), exports);
var flow_run_1 = require("./lib/flow-run/flow-run");
Object.defineProperty(exports, "FlowRun", { enumerable: true, get: function () { return flow_run_1.FlowRun; } });
Object.defineProperty(exports, "RunEnvironment", { enumerable: true, get: function () { return flow_run_1.RunEnvironment; } });
Object.defineProperty(exports, "FlowRetryStrategy", { enumerable: true, get: function () { return flow_run_1.FlowRetryStrategy; } });
tslib_1.__exportStar(require("./lib/flows/dto/create-flow-request"), exports);
var seek_page_1 = require("./lib/common/seek-page");
Object.defineProperty(exports, "SeekPage", { enumerable: true, get: function () { return seek_page_1.SeekPage; } });
var id_generator_1 = require("./lib/common/id-generator");
Object.defineProperty(exports, "apId", { enumerable: true, get: function () { return id_generator_1.apId; } });
Object.defineProperty(exports, "ApId", { enumerable: true, get: function () { return id_generator_1.ApId; } });
Object.defineProperty(exports, "secureApId", { enumerable: true, get: function () { return id_generator_1.secureApId; } });
tslib_1.__exportStar(require("./lib/flows/trigger-events/trigger-events-dto"), exports);
tslib_1.__exportStar(require("./lib/flows/trigger-events/trigger-event"), exports);
tslib_1.__exportStar(require("./lib/flows/sample-data"), exports);
tslib_1.__exportStar(require("./lib/common/base-model"), exports);
tslib_1.__exportStar(require("./lib/flows/folders/folder"), exports);
tslib_1.__exportStar(require("./lib/flows/folders/folder-requests"), exports);
tslib_1.__exportStar(require("./lib/flows/dto/flow-template-request"), exports);
tslib_1.__exportStar(require("./lib/flows"), exports);
tslib_1.__exportStar(require("./lib/flows/dto/list-flows-request"), exports);
tslib_1.__exportStar(require("./lib/project"), exports);
var forms_1 = require("./lib/forms");
Object.defineProperty(exports, "FileResponseInterface", { enumerable: true, get: function () { return forms_1.FileResponseInterface; } });
tslib_1.__exportStar(require("./lib/platform"), exports);
var flow_execution_1 = require("./lib/flow-run/execution/flow-execution");
Object.defineProperty(exports, "isFlowStateTerminal", { enumerable: true, get: function () { return flow_execution_1.isFlowStateTerminal; } });
tslib_1.__exportStar(require("./lib/tag"), exports);
tslib_1.__exportStar(require("./lib/websocket"), exports);
var copilot_1 = require("./lib/copilot");
Object.defineProperty(exports, "GenerateCodeRequest", { enumerable: true, get: function () { return copilot_1.GenerateCodeRequest; } });
Object.defineProperty(exports, "GenerateCodeResponse", { enumerable: true, get: function () { return copilot_1.GenerateCodeResponse; } });
var flow_execution_2 = require("./lib/flow-run/execution/flow-execution");
Object.defineProperty(exports, "FlowError", { enumerable: true, get: function () { return flow_execution_2.FlowError; } });
var flow_execution_3 = require("./lib/flow-run/execution/flow-execution");
Object.defineProperty(exports, "StopResponse", { enumerable: true, get: function () { return flow_execution_3.StopResponse; } });
var flow_execution_4 = require("./lib/flow-run/execution/flow-execution");
Object.defineProperty(exports, "PauseType", { enumerable: true, get: function () { return flow_execution_4.PauseType; } });
Object.defineProperty(exports, "FlowRunStatus", { enumerable: true, get: function () { return flow_execution_4.FlowRunStatus; } });
Object.defineProperty(exports, "FlowRunResponse", { enumerable: true, get: function () { return flow_execution_4.FlowRunResponse; } });
var flow_execution_5 = require("./lib/flow-run/execution/flow-execution");
Object.defineProperty(exports, "DelayPauseMetadata", { enumerable: true, get: function () { return flow_execution_5.DelayPauseMetadata; } });
Object.defineProperty(exports, "PauseMetadata", { enumerable: true, get: function () { return flow_execution_5.PauseMetadata; } });
Object.defineProperty(exports, "WebhookPauseMetadata", { enumerable: true, get: function () { return flow_execution_5.WebhookPauseMetadata; } });
tslib_1.__exportStar(require("./lib/federated-authn"), exports);
var store_entry_1 = require("./lib/store-entry/store-entry");
Object.defineProperty(exports, "STORE_KEY_MAX_LENGTH", { enumerable: true, get: function () { return store_entry_1.STORE_KEY_MAX_LENGTH; } });
var test_flow_run_request_2 = require("./lib/flow-run/test-flow-run-request");
Object.defineProperty(exports, "RetryFlowRequestBody", { enumerable: true, get: function () { return test_flow_run_request_2.RetryFlowRequestBody; } });
tslib_1.__exportStar(require("./lib/flows/dto/flow-template-request"), exports);
tslib_1.__exportStar(require("./lib/support-url"), exports);
tslib_1.__exportStar(require("./lib/license-keys"), exports);
tslib_1.__exportStar(require("./lib/invitations"), exports);
// Look at https://github.com/sinclairzx81/typebox/issues/350
const system_1 = require("@sinclair/typebox/system");
tslib_1.__exportStar(require("./lib/flow-run/execution/flow-execution"), exports);
system_1.TypeSystem.ExactOptionalPropertyTypes = false;
