// ang-jsoneditor/esm2022/jsoneditor/jsoneditoroptions.mjs
/* eslint-disable @typescript-eslint/ban-types */
export class JsonEditorOptions {
    ace;
    ajv;
    /**
     *   {function} onChange  Callback method, triggered
    on change of contents.
    Does not pass the contents itself.
    See also `onChangeJSON` and
    `onChangeText`.
     */
    onChange;
    /**
     *   // {function} onChangeJSON  Callback method, triggered
  //     in modes on change of contents,
  //     passing the changed contents
  //     as JSON.
  //     Only applicable for modes
  //     'tree', 'view', and 'form'.
     */
    onChangeJSON;
    onNodeName;
    onCreateMenu;
    onColorPicker;
    /**
    // {function} onChangeText  Callback method, triggered
    //     in modes on change of contents,
    //     passing the changed contents
    //     as stringified JSON.
     */
    onChangeText;
    /**
     *   {function} onSelectionChange Callback method,
    triggered on node selection change
    Only applicable for modes
    'tree', 'view', and 'form'
     */
    onSelectionChange;
    /**
     *     {function} onTextSelectionChange Callback method,
    triggered on text selection change
    Only applicable for modes
     */
    onTextSelectionChange;
    /**
     *   // {function} onEvent Callback method, triggered
      // when an event occurs in
      // a JSON field or value.
      // Only applicable for
      // modes 'form', 'tree' and
      // 'view'
     */
    onEvent;
    /**
     * // *   {function} onFocus  Callback method, triggered
  //  when the editor comes into focus,
  //  passing an object {type, target},
  //  Applicable for all modes
     */
    onFocus;
    // *   {function} onBlur   Callback method, triggered
    //  when the editor goes out of focus,
    //  passing an object {type, target},
    //  Applicable for all modes
    onBlur;
    /**
     *  // *   {function} onClassName Callback method, triggered
  // when a Node DOM is rendered. Function returns
  // a css class name to be set on a node.
  // Only applicable for
  // modes 'form', 'tree' and
  // 'view'
     */
    onClassName;
    onEditable;
    /**
     *   {function} onError   Callback method, triggered
    when an error occurs
     */
    onError;
    onModeChange;
    onValidate;
    onValidationError;
    enableSort;
    enableTransform;
    escapeUnicode;
    expandAll;
    sortObjectKeys;
    history;
    mode;
    modes;
    name;
    schema;
    search;
    indentation;
    templates;
    theme;
    language;
    languages;
    /**
     * Adds main menu bar - Contains format, sort, transform, search etc. functionality. True
     * by default. Applicable in all types of mode.
     */
    mainMenuBar;
    /**
     * Adds navigation bar to the menu - the navigation bar visualize the current position on
     * the tree structure as well as allows breadcrumbs navigation.
     * True by default.
     * Only applicable when mode is 'tree', 'form' or 'view'.
     */
    navigationBar;
    /**
     * Adds status bar to the bottom of the editor - the status bar shows the cursor position
     * and a count of the selected characters.
     * True by default.
     * Only applicable when mode is 'code' or 'text'.
     */
    statusBar;
    constructor() {
        this.enableSort = true;
        this.enableTransform = true;
        this.escapeUnicode = false;
        this.expandAll = false;
        this.sortObjectKeys = false;
        this.history = true;
        this.mode = 'tree';
        this.search = true;
        this.indentation = 2;
    }
}
