// @activepieces/shared@0.10.107/src/lib/flows/step-run/dto/test-step-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateRunProgressRequest = exports.StepExecutionPath = exports.StepRunResponse = exports.CreateStepRunRequestBody = void 0;
const typebox_1 = require("@sinclair/typebox");
const common_1 = require("../../../common");
const engine_1 = require("../../../engine");
const flow_execution_1 = require("../../../flow-run/execution/flow-execution");
exports.CreateStepRunRequestBody = typebox_1.Type.Object({
    flowVersionId: typebox_1.Type.String(),
    stepName: typebox_1.Type.String(),
    id: typebox_1.Type.String(),
});
exports.StepRunResponse = typebox_1.Type.Object({
    id: typebox_1.Type.String(),
    success: typebox_1.Type.Boolean(),
    output: typebox_1.Type.Unknown(),
    standardError: typebox_1.Type.String(),
    standardOutput: typebox_1.Type.String(),
});
exports.StepExecutionPath = typebox_1.Type.Array(typebox_1.Type.Tuple([typebox_1.Type.String(), typebox_1.Type.Number()]));
exports.UpdateRunProgressRequest = typebox_1.Type.Object({
    runDetails: flow_execution_1.FlowRunResponse,
    runId: typebox_1.Type.String(),
    progressUpdateType: typebox_1.Type.Enum(engine_1.ProgressUpdateType),
    workerHandlerId: (0, common_1.Nullable)(typebox_1.Type.String()),
});
