// @activepieces/shared@0.10.93/src/lib/authentication/model/principal-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformRole = exports.EndpointScope = exports.SERVICE_KEY_SECURITY_OPENAPI = exports.ALL_PRINCIPAL_TYPES = exports.PrincipalType = void 0;
var PrincipalType;
(function (PrincipalType) {
    PrincipalType["USER"] = "USER";
    PrincipalType["WORKER"] = "WORKER";
    PrincipalType["SERVICE"] = "SERVICE";
    PrincipalType["UNKNOWN"] = "UNKNOWN";
    /**
     * @deprecated
     */
    PrincipalType["SUPER_USER"] = "SUPER_USER";
})(PrincipalType || (exports.PrincipalType = PrincipalType = {}));
exports.ALL_PRINCIPAL_TYPES = Object.values(PrincipalType);
exports.SERVICE_KEY_SECURITY_OPENAPI = {
    apiKey: [],
};
var EndpointScope;
(function (EndpointScope) {
    EndpointScope["PLATFORM"] = "PLATFORM";
    EndpointScope["PROJECT"] = "PROJECT";
})(EndpointScope || (exports.EndpointScope = EndpointScope = {}));
var PlatformRole;
(function (PlatformRole) {
    PlatformRole["OWNER"] = "OWNER";
    PlatformRole["MEMBER"] = "MEMBER";
})(PlatformRole || (exports.PlatformRole = PlatformRole = {}));
