// @activepieces/shared@0.10.93/src/lib/project/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectWithLimits = exports.Project = exports.ProjectPlan = exports.ProjectUsage = exports.NotificationStatus = void 0;
const base_model_1 = require("../common/base-model");
const id_generator_1 = require("../common/id-generator");
const typebox_1 = require("@sinclair/typebox");
var NotificationStatus;
(function (NotificationStatus) {
    NotificationStatus["NEVER"] = "NEVER";
    NotificationStatus["ALWAYS"] = "ALWAYS";
})(NotificationStatus || (exports.NotificationStatus = NotificationStatus = {}));
exports.ProjectUsage = typebox_1.Type.Object({
    tasks: typebox_1.Type.Number(),
    teamMembers: typebox_1.Type.Number(),
});
exports.ProjectPlan = typebox_1.Type.Object(Object.assign(Object.assign({}, base_model_1.BaseModelSchema), { projectId: typebox_1.Type.String(), stripeCustomerId: typebox_1.Type.String(), stripeSubscriptionId: (0, base_model_1.Nullable)(typebox_1.Type.String()), subscriptionStartDatetime: typebox_1.Type.String(), flowPlanName: typebox_1.Type.String(), minimumPollingInterval: typebox_1.Type.Number(), connections: typebox_1.Type.Number(), teamMembers: typebox_1.Type.Number(), tasks: typebox_1.Type.Number(), tasksPerDay: (0, base_model_1.Nullable)(typebox_1.Type.Number()) }));
exports.Project = typebox_1.Type.Object(Object.assign(Object.assign({}, base_model_1.BaseModelSchema), { ownerId: typebox_1.Type.String(), displayName: typebox_1.Type.String(), notifyStatus: typebox_1.Type.Enum(NotificationStatus), platformId: id_generator_1.ApId, externalId: typebox_1.Type.Optional(typebox_1.Type.String()) }));
exports.ProjectWithLimits = typebox_1.Type.Composite([
    exports.Project,
    typebox_1.Type.Object({
        usage: exports.ProjectUsage,
        plan: exports.ProjectPlan,
    }),
]);
