// @activepieces/shared@0.10.107/src/lib/flow-run/execution/flow-execution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFailedState = exports.isFlowStateTerminal = exports.FlowRunResponse = exports.FlowError = exports.StopResponse = exports.PauseMetadata = exports.WebhookPauseMetadata = exports.DelayPauseMetadata = exports.PauseType = exports.FlowRunStatus = void 0;
const typebox_1 = require("@sinclair/typebox");
const engine_1 = require("../../engine");
var FlowRunStatus;
(function (FlowRunStatus) {
    FlowRunStatus["FAILED"] = "FAILED";
    FlowRunStatus["QUOTA_EXCEEDED"] = "QUOTA_EXCEEDED";
    FlowRunStatus["INTERNAL_ERROR"] = "INTERNAL_ERROR";
    FlowRunStatus["PAUSED"] = "PAUSED";
    FlowRunStatus["RUNNING"] = "RUNNING";
    FlowRunStatus["STOPPED"] = "STOPPED";
    FlowRunStatus["SUCCEEDED"] = "SUCCEEDED";
    FlowRunStatus["TIMEOUT"] = "TIMEOUT";
})(FlowRunStatus || (exports.FlowRunStatus = FlowRunStatus = {}));
var PauseType;
(function (PauseType) {
    PauseType["DELAY"] = "DELAY";
    PauseType["WEBHOOK"] = "WEBHOOK";
})(PauseType || (exports.PauseType = PauseType = {}));
exports.DelayPauseMetadata = typebox_1.Type.Object({
    type: typebox_1.Type.Literal(PauseType.DELAY),
    resumeDateTime: typebox_1.Type.String(),
    handlerId: typebox_1.Type.Optional(typebox_1.Type.String({})),
    progressUpdateType: typebox_1.Type.Optional(typebox_1.Type.Enum(engine_1.ProgressUpdateType)),
});
exports.WebhookPauseMetadata = typebox_1.Type.Object({
    type: typebox_1.Type.Literal(PauseType.WEBHOOK),
    requestId: typebox_1.Type.String(),
    response: typebox_1.Type.Unknown(),
    handlerId: typebox_1.Type.Optional(typebox_1.Type.String({})),
    progressUpdateType: typebox_1.Type.Optional(typebox_1.Type.Enum(engine_1.ProgressUpdateType)),
});
exports.PauseMetadata = typebox_1.Type.Union([exports.DelayPauseMetadata, exports.WebhookPauseMetadata]);
exports.StopResponse = typebox_1.Type.Object({
    status: typebox_1.Type.Optional(typebox_1.Type.Number()),
    body: typebox_1.Type.Optional(typebox_1.Type.Unknown()),
    headers: typebox_1.Type.Optional(typebox_1.Type.Record(typebox_1.Type.String(), typebox_1.Type.String())),
});
exports.FlowError = typebox_1.Type.Object({
    stepName: typebox_1.Type.String(),
    message: typebox_1.Type.String(),
});
const BaseExecutionResponse = {
    steps: typebox_1.Type.Record(typebox_1.Type.String(), typebox_1.Type.Unknown()),
    duration: typebox_1.Type.Number(),
    tasks: typebox_1.Type.Number(),
    tags: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    error: typebox_1.Type.Optional(exports.FlowError),
    stopResponse: typebox_1.Type.Optional(exports.StopResponse),
};
exports.FlowRunResponse = typebox_1.Type.Union([
    typebox_1.Type.Object(Object.assign(Object.assign({}, BaseExecutionResponse), { status: typebox_1.Type.Literal(FlowRunStatus.PAUSED), pauseMetadata: typebox_1.Type.Optional(exports.PauseMetadata) })),
    typebox_1.Type.Object(Object.assign(Object.assign({}, BaseExecutionResponse), { status: typebox_1.Type.Union([typebox_1.Type.Literal(FlowRunStatus.FAILED),
            typebox_1.Type.Literal(FlowRunStatus.SUCCEEDED),
            typebox_1.Type.Literal(FlowRunStatus.RUNNING),
            typebox_1.Type.Literal(FlowRunStatus.QUOTA_EXCEEDED),
            typebox_1.Type.Literal(FlowRunStatus.TIMEOUT),
            typebox_1.Type.Literal(FlowRunStatus.INTERNAL_ERROR),
            typebox_1.Type.Literal(FlowRunStatus.STOPPED),
        ]) })),
]);
const isFlowStateTerminal = (status) => {
    return status === FlowRunStatus.SUCCEEDED
        || status == FlowRunStatus.STOPPED
        || status === FlowRunStatus.TIMEOUT
        || status === FlowRunStatus.FAILED
        || status === FlowRunStatus.INTERNAL_ERROR
        || status === FlowRunStatus.QUOTA_EXCEEDED;
};
exports.isFlowStateTerminal = isFlowStateTerminal;
const isFailedState = (status) => {
    return status === FlowRunStatus.FAILED
        || status === FlowRunStatus.INTERNAL_ERROR
        || status === FlowRunStatus.QUOTA_EXCEEDED
        || status === FlowRunStatus.TIMEOUT;
};
exports.isFailedState = isFailedState;
