// @activepieces/piece-file-helper@0.0.7/src/lib/actions/create-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFile = void 0;
const tslib_1 = require("tslib");
const pieces_framework_1 = require("@activepieces/pieces-framework");
exports.createFile = (0, pieces_framework_1.createAction)({
    // auth: check https://www.activepieces.com/docs/developers/piece-reference/authentication,
    name: 'createFile',
    displayName: 'Create file',
    description: 'Create file from UTF-8 content',
    props: {
        content: pieces_framework_1.Property.LongText({ displayName: 'Content', required: true }),
        fileName: pieces_framework_1.Property.ShortText({ displayName: 'File name', required: true }),
    },
    run({ propsValue, files }) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fileUrl = yield files.write({
                fileName: propsValue.fileName,
                data: Buffer.from(propsValue.content, 'utf-8'),
            });
            return { fileName: propsValue.fileName, url: fileUrl };
        });
    },
});
