// nx-cloud/lib/utilities/print-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printMessage = void 0;
const fs_1 = require("fs");
const { output } = require('./nx-imports-light');
function printMessage(message) {
    const padding = includePadding() ? '   ' : '';
    if (newTerminalOutput()) {
        process.stdout.write(`${padding}${formatMessage(message)}`);
        output.addNewline();
        output.addNewline();
    }
    else {
        if (runOneCommand()) {
            output.addNewline();
            process.stdout.write(`${padding}${formatMessage(message)}`);
            output.addNewline();
            output.addNewline();
        }
        else {
            process.stdout.write(`${padding}${formatMessage(message)}`);
            output.addNewline();
            output.addNewline();
        }
    }
}
exports.printMessage = printMessage;
const VERSION_REGEX = /(0|[1-9]\d*)(?:\.(0|[1-9]\d*))(?:\.(0|[1-9]\d*))?(?:-.*)?/;
function includePadding() {
    try {
        const packageJson = JSON.parse((0, fs_1.readFileSync)('package.json').toString());
        const deps = Object.assign(Object.assign({}, (packageJson.dependencies || {})), (packageJson.devDependencies || {}));
        const version = deps['nx'].trim().match(VERSION_REGEX);
        const major = +version[1];
        if (major < 18) {
            return true;
        }
        if (major > 18) {
            return false;
        }
        // if version segments are missing (e.g. ^18 or ~15.6) we assume large number
        // to match the logic
        const minor = version[2] ? +version[2] : 9999;
        const patch = version[3] ? +version[3] : 9999;
        if (minor === 0 && patch < 5) {
            return true;
        }
        else {
            return false;
        }
    }
    catch (e) {
        return false;
    }
}
function newTerminalOutput() {
    try {
        require('nx/src/tasks-runner/life-cycles/dynamic-run-many-terminal-output-life-cycle');
        return true;
    }
    catch (e) {
        try {
            require('@nrwl/workspace/src/tasks-runner/life-cycles/dynamic-run-many-terminal-output-life-cycle');
            return true;
        }
        catch (ee) {
            return false;
        }
    }
}
function formatMessage(message) {
    // TODO(Altan): Remove after Nx 15-ish
    // output.dim causes incompatibility with older versions of Nx, so fall back
    // to old method if function is undefined
    if (typeof output.dim === 'function') {
        return output.dim(message);
    }
    else {
        try {
            // Old (pre 13.4/13.5) method
            return output.colors.gray(message);
        }
        catch (e) {
            // Ultra fallback, we should never hit this
            return message;
        }
    }
}
function runOneCommand() {
    return (process.argv.indexOf('run-many') === -1 &&
        process.argv.indexOf('affected') === -1);
}
