// @activepieces/shared@0.10.112/src/lib/project/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateProjectRequestInCommunity = exports.ProjectWithLimits = exports.Project = exports.ProjectPlan = exports.ProjectUsage = exports.NotificationStatus = exports.PiecesFilterType = exports.ListProjectRequestForUserQueryParams = void 0;
const typebox_1 = require("@sinclair/typebox");
const base_model_1 = require("../common/base-model");
const id_generator_1 = require("../common/id-generator");
exports.ListProjectRequestForUserQueryParams = typebox_1.Type.Object({
    cursor: typebox_1.Type.Optional(typebox_1.Type.String()),
    limit: typebox_1.Type.Optional(typebox_1.Type.Number()),
});
var PiecesFilterType;
(function (PiecesFilterType) {
    PiecesFilterType["NONE"] = "NONE";
    PiecesFilterType["ALLOWED"] = "ALLOWED";
})(PiecesFilterType || (exports.PiecesFilterType = PiecesFilterType = {}));
var NotificationStatus;
(function (NotificationStatus) {
    NotificationStatus["NEVER"] = "NEVER";
    NotificationStatus["ALWAYS"] = "ALWAYS";
    NotificationStatus["NEW_ISSUE"] = "NEW_ISSUE";
})(NotificationStatus || (exports.NotificationStatus = NotificationStatus = {}));
exports.ProjectUsage = typebox_1.Type.Object({
    tasks: typebox_1.Type.Number(),
    teamMembers: typebox_1.Type.Number(),
});
exports.ProjectPlan = typebox_1.Type.Object(Object.assign(Object.assign({}, base_model_1.BaseModelSchema), { projectId: typebox_1.Type.String(), name: typebox_1.Type.String(), minimumPollingInterval: typebox_1.Type.Number(), piecesFilterType: typebox_1.Type.Enum(PiecesFilterType), pieces: typebox_1.Type.Array(typebox_1.Type.String()), connections: typebox_1.Type.Number(), teamMembers: typebox_1.Type.Number(), tasks: typebox_1.Type.Number() }));
exports.Project = typebox_1.Type.Object(Object.assign(Object.assign({}, base_model_1.BaseModelSchema), { deleted: (0, base_model_1.Nullable)(typebox_1.Type.String()), ownerId: typebox_1.Type.String(), displayName: typebox_1.Type.String(), notifyStatus: typebox_1.Type.Enum(NotificationStatus), platformId: id_generator_1.ApId, externalId: typebox_1.Type.Optional(typebox_1.Type.String()) }));
exports.ProjectWithLimits = typebox_1.Type.Composite([
    typebox_1.Type.Omit(exports.Project, ['deleted']),
    typebox_1.Type.Object({
        usage: exports.ProjectUsage,
        plan: exports.ProjectPlan,
    }),
]);
exports.UpdateProjectRequestInCommunity = typebox_1.Type.Object({
    notifyStatus: typebox_1.Type.Optional(typebox_1.Type.Enum(NotificationStatus)),
    displayName: typebox_1.Type.Optional(typebox_1.Type.String()),
});
