// @activepieces/shared@0.10.107/src/lib/common/utils/object-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isObject = exports.applyFunctionToValues = exports.applyFunctionToValuesSync = exports.deleteProps = exports.spreadIfDefined = exports.deleteProperties = void 0;
const tslib_1 = require("tslib");
const utils_1 = require("./utils");
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
function deleteProperties(obj, props) {
    const copy = Object.assign({}, obj);
    for (const prop of props) {
        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
        delete copy[prop];
    }
    return copy;
}
exports.deleteProperties = deleteProperties;
const spreadIfDefined = (key, value) => {
    if ((0, utils_1.isNil)(value)) {
        return {};
    }
    return {
        [key]: value,
    };
};
exports.spreadIfDefined = spreadIfDefined;
function deleteProps(obj, prop) {
    const newObj = Object.assign({}, obj);
    for (const p of prop) {
        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
        delete newObj[p];
    }
    return newObj;
}
exports.deleteProps = deleteProps;
function applyFunctionToValuesSync(obj, apply) {
    if ((0, utils_1.isNil)(obj)) {
        return obj;
    }
    else if ((0, utils_1.isString)(obj)) {
        return apply(obj);
    }
    else if (Array.isArray(obj)) {
        for (let i = 0; i < obj.length; ++i) {
            obj[i] = applyFunctionToValuesSync(obj[i], apply);
        }
    }
    else if ((0, exports.isObject)(obj)) {
        const entries = Object.entries(obj);
        for (const entry of entries) {
            const [key, value] = entry;
            obj[key] = applyFunctionToValuesSync(value, apply);
        }
    }
    return apply(obj);
}
exports.applyFunctionToValuesSync = applyFunctionToValuesSync;
function applyFunctionToValues(obj, apply) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if ((0, utils_1.isNil)(obj)) {
            return obj;
        }
        else if ((0, utils_1.isString)(obj)) {
            return yield apply(obj);
        }
        else if (Array.isArray(obj)) {
            for (let i = 0; i < obj.length; ++i) {
                obj[i] = yield applyFunctionToValues(obj[i], apply);
            }
        }
        else if ((0, exports.isObject)(obj)) {
            const entries = Object.entries(obj);
            for (const entry of entries) {
                const [key, value] = entry;
                obj[key] = yield applyFunctionToValues(value, apply);
            }
        }
        return yield apply(obj);
    });
}
exports.applyFunctionToValues = applyFunctionToValues;
const isObject = (obj) => {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
};
exports.isObject = isObject;
