// @activepieces/shared@0.10.93/src/lib/engine/engine-operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EngineResponseStatus = exports.TriggerHookType = exports.EngineOperationType = void 0;
var EngineOperationType;
(function (EngineOperationType) {
    EngineOperationType["EXTRACT_PIECE_METADATA"] = "EXTRACT_PIECE_METADATA";
    EngineOperationType["EXECUTE_STEP"] = "EXECUTE_STEP";
    EngineOperationType["EXECUTE_TEST_FLOW"] = "EXECUTE_TEST_FLOW";
    EngineOperationType["EXECUTE_FLOW"] = "EXECUTE_FLOW";
    EngineOperationType["EXECUTE_PROPERTY"] = "EXECUTE_PROPERTY";
    EngineOperationType["EXECUTE_TRIGGER_HOOK"] = "EXECUTE_TRIGGER_HOOK";
    EngineOperationType["EXECUTE_VALIDATE_AUTH"] = "EXECUTE_VALIDATE_AUTH";
})(EngineOperationType || (exports.EngineOperationType = EngineOperationType = {}));
var TriggerHookType;
(function (TriggerHookType) {
    TriggerHookType["ON_ENABLE"] = "ON_ENABLE";
    TriggerHookType["ON_DISABLE"] = "ON_DISABLE";
    TriggerHookType["HANDSHAKE"] = "HANDSHAKE";
    TriggerHookType["RENEW"] = "RENEW";
    TriggerHookType["RUN"] = "RUN";
    TriggerHookType["TEST"] = "TEST";
})(TriggerHookType || (exports.TriggerHookType = TriggerHookType = {}));
var EngineResponseStatus;
(function (EngineResponseStatus) {
    EngineResponseStatus["OK"] = "OK";
    EngineResponseStatus["ERROR"] = "ERROR";
    EngineResponseStatus["TIMEOUT"] = "TIMEOUT";
})(EngineResponseStatus || (exports.EngineResponseStatus = EngineResponseStatus = {}));
