// @activepieces/shared@0.10.93/src/lib/flag/flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApFlagId = exports.ApEdition = exports.ApEnvironment = void 0;
var ApEnvironment;
(function (ApEnvironment) {
    ApEnvironment["PRODUCTION"] = "prod";
    ApEnvironment["DEVELOPMENT"] = "dev";
    ApEnvironment["TESTING"] = "test";
})(ApEnvironment || (exports.ApEnvironment = ApEnvironment = {}));
var ApEdition;
(function (ApEdition) {
    ApEdition["COMMUNITY"] = "ce";
    ApEdition["ENTERPRISE"] = "ee";
    ApEdition["CLOUD"] = "cloud";
})(ApEdition || (exports.ApEdition = ApEdition = {}));
var ApFlagId;
(function (ApFlagId) {
    ApFlagId["BILLING_ENABLED"] = "BILLING_ENABLED";
    ApFlagId["CLOUD_AUTH_ENABLED"] = "CLOUD_AUTH_ENABLED";
    ApFlagId["COPILOT_ENABLED"] = "COPILOT_ENABLED";
    ApFlagId["CURRENT_VERSION"] = "CURRENT_VERSION";
    ApFlagId["EDITION"] = "EDITION";
    ApFlagId["EMAIL_AUTH_ENABLED"] = "EMAIL_AUTH_ENABLED";
    ApFlagId["ENVIRONMENT"] = "ENVIRONMENT";
    ApFlagId["FRONTEND_URL"] = "FRONTEND_URL";
    ApFlagId["LATEST_VERSION"] = "LATEST_VERSION";
    ApFlagId["OWN_AUTH2_ENABLED"] = "OWN_AUTH2_ENABLED";
    ApFlagId["PLATFORM_CREATED"] = "PLATFORM_CREATED";
    ApFlagId["PRIVACY_POLICY_URL"] = "PRIVACY_POLICY_URL";
    ApFlagId["PRIVATE_PIECES_ENABLED"] = "PRIVATE_PIECES_ENABLED";
    ApFlagId["PROJECT_MEMBERS_ENABLED"] = "PROJECT_MEMBERS_ENABLED";
    ApFlagId["SANDBOX_RUN_TIME_SECONDS"] = "SANDBOX_RUN_TIME_SECONDS";
    ApFlagId["SHOW_ACTIVITY_LOG"] = "SHOW_ACTIVITY_LOG";
    ApFlagId["SHOW_BILLING"] = "SHOW_BILLING";
    ApFlagId["SHOW_BLOG_GUIDE"] = "SHOW_BLOG_GUIDE";
    ApFlagId["SHOW_COMMUNITY_PIECES"] = "SHOW_COMMUNITY_PIECES";
    ApFlagId["SHOW_COMMUNITY"] = "SHOW_COMMUNITY";
    ApFlagId["SHOW_COPILOT"] = "SHOW_COPILOT";
    ApFlagId["SHOW_DOCS"] = "SHOW_DOCS";
    ApFlagId["SHOW_GIT_SYNC"] = "SHOW_GIT_SYNC";
    ApFlagId["SHOW_PLATFORM_DEMO"] = "SHOW_PLATFORM_DEMO";
    ApFlagId["SHOW_POWERED_BY_AP"] = "SHOW_POWERED_BY_AP";
    ApFlagId["SHOW_SIGN_UP_LINK"] = "SHOW_SIGN_UP_LINK";
    ApFlagId["SIGN_UP_ENABLED"] = "SIGN_UP_ENABLED";
    ApFlagId["SUPPORTED_APP_WEBHOOKS"] = "SUPPORTED_APP_WEBHOOKS";
    ApFlagId["TELEMETRY_ENABLED"] = "TELEMETRY_ENABLED";
    ApFlagId["TEMPLATES_PROJECT_ID"] = "TEMPLATES_PROJECT_ID";
    ApFlagId["TERMS_OF_SERVICE_URL"] = "TERMS_OF_SERVICE_URL";
    ApFlagId["THEME"] = "THEME";
    ApFlagId["THIRD_PARTY_AUTH_PROVIDER_REDIRECT_URL"] = "THIRD_PARTY_AUTH_PROVIDER_REDIRECT_URL";
    ApFlagId["THIRD_PARTY_AUTH_PROVIDERS_TO_SHOW_MAP"] = "THIRD_PARTY_AUTH_PROVIDERS_TO_SHOW_MAP";
    ApFlagId["USER_CREATED"] = "USER_CREATED";
    ApFlagId["WEBHOOK_URL_PREFIX"] = "WEBHOOK_URL_PREFIX";
})(ApFlagId || (exports.ApFlagId = ApFlagId = {}));
