// @activepieces/shared@0.10.93/src/lib/platform/platform.request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminAddPlatformRequestBody = exports.UpdatePlatformRequestBody = void 0;
const typebox_1 = require("@sinclair/typebox");
const platform_model_1 = require("./platform.model");
const federated_authn_1 = require("../federated-authn");
const common_1 = require("../common");
const id_generator_1 = require("../common/id-generator");
exports.UpdatePlatformRequestBody = typebox_1.Type.Object({
    name: typebox_1.Type.Optional(typebox_1.Type.String()),
    primaryColor: typebox_1.Type.Optional(typebox_1.Type.String()),
    logoIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    fullLogoUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    favIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    filteredPieceNames: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    filteredPieceBehavior: typebox_1.Type.Optional(typebox_1.Type.Enum(platform_model_1.FilteredPieceBehavior)),
    smtpHost: typebox_1.Type.Optional(typebox_1.Type.String()),
    smtpPort: typebox_1.Type.Optional(typebox_1.Type.Number()),
    smtpUser: typebox_1.Type.Optional(typebox_1.Type.String()),
    smtpPassword: typebox_1.Type.Optional(typebox_1.Type.String()),
    smtpSenderEmail: typebox_1.Type.Optional(typebox_1.Type.String()),
    smtpUseSSL: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    privacyPolicyUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    federatedAuthProviders: typebox_1.Type.Optional(federated_authn_1.FederatedAuthnProviderConfig),
    termsOfServiceUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    cloudAuthEnabled: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    emailAuthEnabled: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    allowedAuthDomains: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String())),
    enforceAllowedAuthDomains: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    defaultLocale: typebox_1.Type.Optional(typebox_1.Type.Enum(common_1.LocalesEnum)),
});
exports.AdminAddPlatformRequestBody = typebox_1.Type.Object({
    userId: id_generator_1.ApId,
    projectId: id_generator_1.ApId,
    name: typebox_1.Type.String(),
    primaryColor: typebox_1.Type.Optional(typebox_1.Type.String()),
    logoIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    fullLogoUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
    favIconUrl: typebox_1.Type.Optional(typebox_1.Type.String()),
});
